### PR TITLE
Asynchronous code for (real) execute method

### PR DESCRIPTION
**Pull request by krinsman**
Added what the asynchronous version of the `execute` method might look like, at least inasmuch as the mental model exhibited in the dummies was correct.

All that I have to do now is modify the 3 methods depending on `execute` to be coroutines and then try to test it locally to see if there's blocking.

### CODE REVIEW

**Comment by rcthomas** [comment]
So two thoughts.  Could we use `exec` instead of the `shell` method?  And I see in the docs that "All other keyword arguments are passed to subprocess.Popen without interpretation, except for bufsize, universal_newlines and shell, which should not be specified at all" which suggests to me that passing `env` should work.

**Comment by krinsman** [comment]
@rcthomas Let me double-check about `shell` vs. `exec`.

The thing about `exec`, if I remember correctly at the moment, is that it requires every word in the command to be passed as a separate argument. In the original `Popen` one can just pass a list of strings, but in `exec` every single string has to be passed as a separate argument. It might be possible to make that work, but it would be a major headache. Especially since we're actually using the shell, the syntax for shell seems to match better.

**EDIT:** [It doesn't seem that](https://docs.python.org/3/library/asyncio-subprocess.html#create-a-subprocess-high-level-api-using-process) either `exec` or `shell` have an `env` parameter. The major difference seems to be the syntax of how to submit a command, for which `shell` is much better.

**Comment by rcthomas** [comment]
OK I think this is not critical; the old code is using `shell=True` for various reasons and I think resolving those would be beyond the scope of this PR.  Does `env` look like it'll work?

**Comment by krinsman** [comment]
@rcthomas Yes, it looks like `env` will work. This is good, since it simplifies the code substantially.

File: sshspawner/sshspawner.py [hunk 1]
```diff
@@ -98,62 +98,75 @@ def get_gsi_key(self):
...
+    self.log.debug("command: {}".format(command))
+    
+    try:
+        proc = await asyncio.create_subprocess_shell(command, 
+                                                    stdout=asyncio.subprocess.PIPE, 
+                                                    stderr=asyncio.subprocess.PIPE)
         try:
-            stdout, stderr = proc.communicate(timeout=10)
-        except TimeoutExpired:
+            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=10)
```
Comment (rcthomas) [inline sshspawner/sshspawner.py]: How about let's try without the timeout here.

File: sshspawner/sshspawner.py [hunk 2]
```diff
@@ -98,62 +98,75 @@ def get_gsi_key(self):
...
+        except asyncio.TimeoutError:
```
Comment (rcthomas) [inline sshspawner/sshspawner.py:148]: Similarly I think we can take out the try/except block.

Comment (krinsman) [reply sshspawner/sshspawner.py:148]: I don't think I understand; is the timeout blocking? How can the `communicate` call still fail gracefully without the try/except block?

**Comment by krinsman** [comment]
Just realized I forgot the most obvious thing: `import asyncio` at the top of the file. Will fix this soon.

**Pull request closed without merging**